### PR TITLE
Column type independent materialize()

### LINKED ISSUE #695: Update Projection to support all encoded column types
The `Projection` class does only support the three old column types at the moment. See `Projection::_evaluate_expression()`

### PR DESCRIPTION
**Pull request by mrzzzrm**
Fixes #695

### CODE REVIEW

File: src/lib/storage/column_iterables.hpp [hunk 1]
```diff
@@ -80,6 +80,45 @@ class ColumnIterable {
     });
   }
 
+  /**
+   * Materialize all values in this Iterable.
+   * Fallback implementation. Derived Iterables can implement their own, faster version.
```
Comment (mrks) [inline src/lib/storage/column_iterables.hpp]: Should we add a PerformanceWarning for this?

Comment (mrzzzrm) [reply src/lib/storage/column_iterables.hpp]: No, it's probably the best possible implementation for anything but the ValueColumn

**Comment by mrzzzrm** [comment]
Added tests, ready from my side.

File: src/test/storage/encoding_test.hpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    const auto table = load_table(path, max_chunk_size);
+    if (GetParam().encoding_type != EncodingType::Unencoded) {
+      ChunkEncoder::encode_all_chunks(table, GetParam());
+    }
+    return table;
+  }
+};
+
+// clang-format off
+#define INSTANCIATE_ENCODING_TYPE_TESTS(Test) \
```
Comment (mrzzzrm) [inline src/test/storage/encoding_test.hpp]: @mjendruk does this cover all Encoding-VectorCompression combinations?

Comment (mjendruk) [reply src/test/storage/encoding_test.hpp]: Yes

Comment (mjendruk) [reply src/test/storage/encoding_test.hpp]: Not sure if the macro is really necessary. We could simply store `::testing::Values(...)` somewhere.

File: src/test/storage/encoding_test.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+#include "storage/table.hpp"
+#include "utils/load_table.hpp"
+
+namespace opossum {
+
+class EncodingTest : public ::testing::TestWithParam<ColumnEncodingSpec> {
+ public:
+  std::shared_ptr<Table> load_table_with_encoding(const std::string& path, ChunkOffset max_chunk_size) {
+    const auto table = load_table(path, max_chunk_size);
+    if (GetParam().encoding_type != EncodingType::Unencoded) {
```
Comment (mjendruk) [inline src/test/storage/encoding_test.hpp]: ChunkEncoder can also handle `EncodingType::Unencoded`.

File: src/test/storage/encoding_test.hpp (hunk 2 above)
Comment (mrzzzrm) [reply src/test/storage/encoding_test.hpp]: done

File: src/test/storage/encoding_test.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+// clang-format off
+#define INSTANCIATE_ENCODING_TYPE_TESTS(Test) \
+  INSTANTIATE_TEST_CASE_P(Instances##Test, Test, ::testing::Values( \
+    ColumnEncodingSpec(EncodingType::Unencoded), \
+    ColumnEncodingSpec(EncodingType::DeprecatedDictionary), \
+    ColumnEncodingSpec(EncodingType::Dictionary, VectorCompressionType::FixedSizeByteAligned), \
+    ColumnEncodingSpec(EncodingType::Dictionary, VectorCompressionType::SimdBp128), \
+    ColumnEncodingSpec(EncodingType::Dictionary, VectorCompressionType::SimdBp128), \
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/test/storage/encoding_test.hpp]: quite so

File: src/test/storage/encoding_test.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include "gtest/gtest.h"
+
+#include "storage/chunk_encoder.hpp"
+#include "storage/encoding_type.hpp"
+#include "storage/table.hpp"
+#include "utils/load_table.hpp"
+
+namespace opossum {
+
+class EncodingTest : public ::testing::TestWithParam<ColumnEncodingSpec> {
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/test/storage/encoding_test.hpp]: nope

File: src/lib/storage/materialize.hpp [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace opossum {
+
+/**
+ * Materialization convenience functions. Can't be put into base_column.hpp because that would lead to circular
+ * includes.
+ *
+ * Use like:
+ *
+ * pmr_concurrent_vector<std::optional<T>> values_and_nulls;
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/materialize.hpp]: yes

File: src/lib/storage/materialize.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+ * values_and_nulls.reserve(chunk->size()); // Optional
+ * materialize_values_and_nulls(*chunk->get_column(expression->column_id()), values_and_nulls);
+ * return values_and_nulls;
+ */
+
+// Materialize the values in the Column
+template <template <typename> typename Container, typename T>
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/materialize.hpp]: clever

File: src/lib/storage/column_iterables.hpp [hunk 8]
```diff
@@ -80,6 +80,45 @@ class ColumnIterable {
...
+  /**
+   * Materialize all values in this Iterable as std::optional<ValueType>. std::nullopt if value is NULL.
+   * Fallback implementation. Derived Iterables can implement their own, faster version.
+   * @param container   The container the values of this Iterable will be push_back()ed into
+   */
+  template <typename Container>
+  void materialize_values_and_nulls(Container& container) const {
+    for_each([&](const auto& iter) {
+      if (iter.is_null()) {
+        container.push_back(std::nullopt);
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/column_iterables.hpp]: if you say so. Then Projection probably just got faster as well, lots of branching in there.

File: src/lib/storage/column_iterables.hpp [hunk 9]
```diff
@@ -80,6 +80,45 @@ class ColumnIterable {
...
   }
 
+  /**
+   * Materialize all values in this Iterable.
+   * Fallback implementation. Derived Iterables can implement their own, faster version.
+   * @param container   The container the values of this Iterable will be push_back()ed into.
+   */
+  template <typename Container>
+  void materialize_values(Container& container) const {
+    for_each([&](const auto& iter) {
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/column_iterables.hpp]: We can't have a ValueType in AnyColumnIterable, for the reason we discussed on telegram: https://godbolt.org/g/uj4XYA

I wouldn't do it.

File: src/lib/storage/materialize.hpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+void materialize_values_and_nulls(const BaseColumn& column, Container<std::optional<T>>& container) {
+  resolve_column_type<T>(column, [&](const auto& column) {
+    create_iterable_from_column<T>(column).materialize_values_and_nulls(container);
+  });
+}
+
+// Materialize the nulls in the Column
+template <typename ColumnValueType, typename Container>
+void materialize_nulls(const BaseColumn& column, Container& container) {
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/storage/materialize.hpp]: I think implicit conversion is not something you would want to forbid here. Remember, its a general purpose template lib ;)

**Comment by mrzzzrm** [comment]
Covered that review, I think

File: src/lib/storage/column_iterables.hpp [hunk 11]
```diff
@@ -80,6 +80,49 @@ class ColumnIterable {
...
   }
 
+  /**
+   * @defgroup Function for the materialization of values and nulls
+   * @{
+   */
+
+  /**
+   * Materialize all values in this iterable.
+   * Fallback implementation. Derived iterables can implement their own, faster version.
```
Comment (mjendruk) [inline src/lib/storage/column_iterables.hpp]: Could you add this line to the group instead? 

`Fallback implementation. Derived iterables can implement their own, faster version.`

Perhaps:
`The following implementations may be overridden by derived classes`

Comment (mrzzzrm) [reply src/lib/storage/column_iterables.hpp]: done

File: src/lib/storage/materialize.hpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace opossum {
+
+/**
+ * Materialization convenience functions. Can't be put into base_column.hpp because that would lead to circular
+ * includes.
+ *
+ * Use like:
+ *
+ * pmr_vector<std::optional<T>> values_and_nulls;
```
Comment (mjendruk) [inline src/lib/storage/materialize.hpp]: `std::pair<bool, T>` ;)

Also, I would indent the example by two spaces. This is how I did it everywhere else.

Comment (mrzzzrm) [reply src/lib/storage/materialize.hpp]: done

File: src/lib/storage/materialize.hpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+template <typename Container>
+void materialize_values_and_nulls(const BaseColumn& column, Container& container) {
+  using ContainerValueType = typename Container::value_type::second_type;
+  resolve_column_type<ContainerValueType>(column, [&](const auto& column) {
+    create_iterable_from_column<ContainerValueType>(column).materialize_values_and_nulls(container);
+  });
+}
+
+// Materialize the nulls in the Column
+template <typename ColumnValueType, typename BoolContainer>
```
Comment (mjendruk) [inline src/lib/storage/materialize.hpp]: `BoolContainer` is confusing. You don’t call the container for values and nulls `ValueBoolContainer` either.

File: src/lib/storage/reference_column.hpp [hunk 14]
```diff
@@ -28,47 +28,6 @@ class ReferenceColumn : public BaseColumn {
 
   void append(const AllTypeVariant&) override;
 
-  // return generated vector of all values (or nulls)
-  template <typename T>
-  const pmr_concurrent_vector<std::optional<T>> materialize_values() const {
```
Comment (mjendruk) [inline src/lib/storage/reference_column.hpp]: Yes!!

File: src/lib/storage/materialize.hpp (hunk 13 above)
Comment (mrzzzrm) [reply src/lib/storage/materialize.hpp]: for you

File: src/test/storage/materialize_test.cpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+  const auto nulls =
+      materialize_nulls_to_vector<float>(*_data_table_with_nulls->get_chunk(ChunkID(0))->get_column(ColumnID(1)));
+  auto expected = std::vector<bool>{false, true};
+
+  EXPECT_EQ(expected, nulls);
+}
+
+INSTANTIATE_TEST_CASE_P(MaterializeTestInstances, MaterializeTest,
+                        ::testing::ValuesIn(std::begin(all_column_encoding_specs),
+                                            std::end(all_column_encoding_specs)), ); // NOLINT
```
Comment (mjendruk) [inline src/test/storage/materialize_test.cpp]: Could you add `_` here after `std::end(all_column_encoding_specs)), `? This is what we do in other places.

File: src/lib/storage/column_iterables.hpp [hunk 16]
```diff
@@ -80,6 +80,47 @@ class ColumnIterable {
     });
   }
 
+  /**
+   * @defgroup Function for the materialization of values and nulls. The following implementations may be overridden by
```
Comment (mjendruk) [inline src/lib/storage/column_iterables.hpp]: I’m sorry Moritz. Could you put the last sentence on a new line, please? What comes after `@defgroup` is the name of the group. The rest is just additional information.

File: src/test/storage/materialize_test.cpp (hunk 15 above)
Comment (mrzzzrm) [reply src/test/storage/materialize_test.cpp]: According to Ctrl+Shift+F we don't ;)

File: src/lib/storage/column_iterables.hpp (hunk 16 above)
Comment (mrzzzrm) [reply src/lib/storage/column_iterables.hpp]: sure

File: src/test/storage/materialize_test.cpp (hunk 15 above)
Comment (mjendruk) [reply src/test/storage/materialize_test.cpp]: Yes we do. 

https://github.com/hyrise/hyrise/blob/75d8d8d7e6c58791abb0f7c8e07a3207f76e99df/src/lib/all_type_variant.hpp#L62

**Pull request merged**